02.recognize_digits/index.html

```diff
@@ -315,48 +321,100 @@
-
+lists = []
 def event_handler(event):
+    if isinstance(event, fluid.EndStepEvent):
+        if event.step % 100 == 0:
+            # event.metrics maps with train program return arguments.
+            # event.metrics[0] will yeild avg_cost and event.metrics[1] will yeild acc in this example.
+            print "Pass %d, Batch %d, Cost %f" % (
+                event.step, event.epoch, event.metrics[0])
+
     if isinstance(event, fluid.EndEpochEvent):
         avg_cost, acc = trainer.test(
             reader=test_reader, feed_order=['img', 'label'])
-        print("avg_cost: %s, acc: %s" % (avg_cost, acc))
+
+        print("Test with Epoch %d, avg_cost: %s, acc: %s" % (event.epoch, avg_cost, acc))
+
+        # save parameters
+        trainer.save_params(params_dirname)
+        lists.append((event.epoch, avg_cost, acc))
+```
+
+`event_handler_plot` is used to plot a figure like below：
+
+![png](./image/train_and_test.png)
+
+```python
+from paddle.v2.plot import Ploter
+
+train_title = "Train cost"
+test_title = "Test cost"
+cost_ploter = Ploter(train_title, test_title)
+step = 0
+lists = []
+
+# event_handler to plot a figure
+def event_handler_plot(event):
+    global step
+    if isinstance(event, fluid.EndStepEvent):
+        if step % 100 == 0:
+            # event.metrics maps with train program return arguments.
+            # event.metrics[0] will yeild avg_cost and event.metrics[1] will yeild acc in this example.
+            cost_ploter.append(train_title, step, event.metrics[0])
+            cost_ploter.plot()
+        step += 1
+    if isinstance(event, fluid.EndEpochEvent):
+        # save parameters
         trainer.save_params(params_dirname)
+
+        avg_cost, acc = trainer.test(
+            reader=test_reader, feed_order=['img', 'label'])
+        cost_ploter.append(test_title, step, avg_cost)
+        lists.append((event.epoch, avg_cost, acc))
 ```
 
 Now that we setup the event_handler and the reader, we can start training the model. `feed_order` is used to map the data dict to the train_program
 ```python
 # Train the model now
 trainer.train(
-    num_epochs=1,
-    event_handler=event_handler,
+    num_epochs=5,
+    event_handler=event_handler_plot,
     reader=train_reader,
     feed_order=['img', 'label'])
 ```
 
 During training, `trainer.train` invokes `event_handler` for certain events. This gives us a chance to print the training progress.
 
- ```
- # Pass 0, Batch 0, Cost 2.780790, {'classification_error_evaluator': 0.9453125}
- # Pass 0, Batch 100, Cost 0.635356, {'classification_error_evaluator': 0.2109375}
- # Pass 0, Batch 200, Cost 0.326094, {'classification_error_evaluator': 0.1328125}
- # Pass 0, Batch 300, Cost 0.361920, {'classification_error_evaluator': 0.1015625}
- # Pass 0, Batch 400, Cost 0.410101, {'classification_error_evaluator': 0.125}
- # Test with Pass 0, Cost 0.326659, {'classification_error_evaluator': 0.09470000118017197}
- ```
+```
+Pass 0, Batch 0, Cost 0.125650
```

Comment:
Are these values ('Cost' values) consistent across different runs of the same program?

Reply:
They don't need to be the same. The print is only for reference to show that the cost will decrease.